Add support for more password types in SQLIdResolver

andyboeh opened the pull request:
This is a follow-up to #586, but since more than a year has passed, I'm creating a new PR here.

I added support for MD5Crypt (i.e. $1$) and SHA512-Crypt (i.e. $6$) password schemes using the crypt.crypt() module, including test cases. Two new users were added to testuser.sqlite, one containing an MD5Crypt hash and one containing an SHA512-Crypt hash. I hope that I updated all other test cases as needed.

Further, I added hashing support for MD5Crypt and SHA512Crypt. I didn't check whether there is documentation that needs to be udpated, though.
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/privacyidea/privacyidea/pull/1040%23issuecomment-386049674%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1040%23discussion_r186070706%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1040%23discussion_r186113242%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1040%23issuecomment-386747311%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1040%23issuecomment-386863926%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1040%23issuecomment-386866879%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1040%23issuecomment-386904089%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/privacyidea/privacyidea/pull/1040%23issuecomment-386049674%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%23%20%5BCodecov%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1040%3Fsrc%3Dpr%26el%3Dh1%29%20Report%5Cn%3E%20Merging%20%5B%231040%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1040%3Fsrc%3Dpr%26el%3Ddesc%29%20into%20%5Bmaster%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/commit/c8cb96c74b97815339914a92fbc94b43a97d788a%3Fsrc%3Dpr%26el%3Ddesc%29%20will%20%2A%2Aincrease%2A%2A%20coverage%20by%20%60%3C.01%25%60.%5Cn%3E%20The%20diff%20coverage%20is%20%60100%25%60.%5Cn%5Cn%5B%21%5BImpacted%20file%20tree%20graph%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1040/graphs/tree.svg%3Fsrc%3Dpr%26token%3D7nHLZki60B%26width%3D650%26height%3D150%29%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1040%3Fsrc%3Dpr%26el%3Dtree%29%5Cn%5Cn%60%60%60diff%5Cn%40%40%20%20%20%20%20%20%20%20%20%20%20%20Coverage%20Diff%20%20%20%20%20%20%20%20%20%20%20%20%20%40%40%5Cn%23%23%20%20%20%20%20%20%20%20%20%20%20master%20%20%20%20%231040%20%20%20%20%20%20%2B/-%20%20%20%23%23%5Cn%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%5Cn%2B%20Coverage%20%20%2095.41%25%20%20%2095.41%25%20%20%20%2B%3C.01%25%20%20%20%20%20%5Cn%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%5Cn%20%20Files%20%20%20%20%20%20%20%20%20131%20%20%20%20%20%20131%20%20%20%20%20%20%20%20%20%20%20%20%20%20%5Cn%20%20Lines%20%20%20%20%20%20%2016199%20%20%20%2016203%20%20%20%20%20%20%20%2B4%20%20%20%20%20%5Cn%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%5Cn%2B%20Hits%20%20%20%20%20%20%20%2015456%20%20%20%2015460%20%20%20%20%20%20%20%2B4%20%20%20%20%20%5Cn%20%20Misses%20%20%20%20%20%20%20%20743%20%20%20%20%20%20743%5Cn%60%60%60%5Cn%5Cn%5Cn%7C%20%5BImpacted%20Files%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1040%3Fsrc%3Dpr%26el%3Dtree%29%20%7C%20Coverage%20%5Cu0394%20%7C%20%7C%5Cn%7C---%7C---%7C---%7C%5Cn%7C%20%5Bprivacyidea/lib/utils.py%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1040/diff%3Fsrc%3Dpr%26el%3Dtree%23diff-cHJpdmFjeWlkZWEvbGliL3V0aWxzLnB5%29%20%7C%20%6096.69%25%20%3C100%25%3E%20%28%2B0.01%25%29%60%20%7C%20%3Aarrow_up%3A%20%7C%5Cn%7C%20%5Bprivacyidea/lib/resolvers/SQLIdResolver.py%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1040/diff%3Fsrc%3Dpr%26el%3Dtree%23diff-cHJpdmFjeWlkZWEvbGliL3Jlc29sdmVycy9TUUxJZFJlc29sdmVyLnB5%29%20%7C%20%6097.44%25%20%3C100%25%3E%20%28%2B0.01%25%29%60%20%7C%20%3Aarrow_up%3A%20%7C%5Cn%5Cn------%5Cn%5Cn%5BContinue%20to%20review%20full%20report%20at%20Codecov%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1040%3Fsrc%3Dpr%26el%3Dcontinue%29.%5Cn%3E%20%2A%2ALegend%2A%2A%20-%20%5BClick%20here%20to%20learn%20more%5D%28https%3A//docs.codecov.io/docs/codecov-delta%29%5Cn%3E%20%60%5Cu0394%20%3D%20absolute%20%3Crelative%3E%20%28impact%29%60%2C%20%60%5Cu00f8%20%3D%20not%20affected%60%2C%20%60%3F%20%3D%20missing%20data%60%5Cn%3E%20Powered%20by%20%5BCodecov%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1040%3Fsrc%3Dpr%26el%3Dfooter%29.%20Last%20update%20%5Bc8cb96c...6f5d4f1%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1040%3Fsrc%3Dpr%26el%3Dlastupdated%29.%20Read%20the%20%5Bcomment%20docs%5D%28https%3A//docs.codecov.io/docs/pull-request-comments%29.%5Cn%22%2C%20%22created_at%22%3A%20%222018-05-02T17%3A05%3A20Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars0.githubusercontent.com/u/8655789%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/codecov-io%22%7D%7D%2C%20%7B%22body%22%3A%20%22Thanks%20for%20the%20adaptions%2C%20they%20look%20good.%5Cr%5CnI%20just%20realized%2C%20that%20your%20PR%20is%20based%20on%20version%202.17.%20Would%20you%20mind%20rebasing%20it%20to%20the%20current%20master%3F%22%2C%20%22created_at%22%3A%20%222018-05-04T22%3A06%3A54Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars1.githubusercontent.com/u/1908620%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%27m%20not%20very%20proficient%20in%20using%20git%2C%20but%20I%20know%20what%20you%20would%20like%20to%20achieve%20-%20my%20problem%20is%20that%20I%20started%20with%20master%20and%20not%20a%20feature%20branch%20and%20that%20I%20did%20a%20merge%20instead%20of%20a%20rebase%20in%20the%20first%20place..%22%2C%20%22created_at%22%3A%20%222018-05-06T08%3A46%3A06Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/3099753%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/andyboeh%22%7D%7D%2C%20%7B%22body%22%3A%20%22OK%2C%20I%20kind%20of%20fixed%20it%20locally%2C%20but%20if%20I%20push%20it%20now%2C%20I%27m%20pretty%20sure%20it%27s%20going%20to%20destroy%20the%20PR%2C%20since%20I%20moved%20the%20relevant%20commits%20to%20a%20feature%20branch%20and%20cleaned-up%20my%20master%20branch.%20Should%20I%20create%20a%20new%20PR%2C%20referencing%20this%20one%3F%22%2C%20%22created_at%22%3A%20%222018-05-06T09%3A40%3A48Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/3099753%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/andyboeh%22%7D%7D%2C%20%7B%22body%22%3A%20%22Thanks%20for%20all%20your%20effort.%20I%20will%20probably%20take%20a%20look%20at%20it%20again%20tomorrow%20and%20maybe%20can%20fix%20this.%22%2C%20%22created_at%22%3A%20%222018-05-06T18%3A57%3A59Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars1.githubusercontent.com/u/1908620%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%206f5d4f1e1f8a1372ddfbe10f9b599000a0b80c71%20privacyidea/lib/utils.py%209%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/privacyidea/privacyidea/pull/1040%23discussion_r186070706%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22To%20be%20able%20to%20choose%20these%20hashtypes%20%28MD5CRYPT%2C%20SHA512CRYPT%29%20in%20a%20configurable%20SQL%20Resolver%2C%20you%20need%20to%20add%20this%20to%20the%20WebUI%20dropdown%20here%3A%5Cr%5Cn%5Cr%5Cnhttps%3A//github.com/privacyidea/privacyidea/blob/master/privacyidea/static/components/config/controllers/configControllers.js%23L990%5Cr%5Cn%5Cr%5Cn%22%2C%20%22created_at%22%3A%20%222018-05-04T12%3A47%3A04Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars1.githubusercontent.com/u/1908620%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20privacyidea/lib/utils.py%3AL849-868%22%7D%2C%20%22Pull%206f5d4f1e1f8a1372ddfbe10f9b599000a0b80c71%20privacyidea/lib/utils.py%208%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/privacyidea/privacyidea/pull/1040%23discussion_r186113242%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22You%20are%20right%2C%20up%20to%2016%20bytes%20are%20supported%20by%20crypt%283%29%2C%20while%20I%20had%20exactly%208%20bytes%20in%20mind.%20I%27ll%20double%20that.%22%2C%20%22created_at%22%3A%20%222018-05-04T15%3A10%3A12Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/3099753%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/andyboeh%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20privacyidea/lib/utils.py%3AL849-868%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'><img src='http://www.codereviewhub.com/site/github-bar.png' height=40></a>
- [ ] <a href='#crh-comment-General Comment'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/privacyidea/privacyidea/pull/1040#issuecomment-386049674'>General Comment</a></b>
- <a href='https://github.com/codecov-io'><img border=0 src='https://avatars0.githubusercontent.com/u/8655789?v=4' height=16 width=16></a> # [Codecov](https://codecov.io/gh/privacyidea/privacyidea/pull/1040?src=pr&el=h1) Report
> Merging [#1040](https://codecov.io/gh/privacyidea/privacyidea/pull/1040?src=pr&el=desc) into [master](https://codecov.io/gh/privacyidea/privacyidea/commit/c8cb96c74b97815339914a92fbc94b43a97d788a?src=pr&el=desc) will **increase** coverage by `<.01%`.
> The diff coverage is `100%`.
[![Impacted file tree graph](https://codecov.io/gh/privacyidea/privacyidea/pull/1040/graphs/tree.svg?src=pr&token=7nHLZki60B&width=650&height=150)](https://codecov.io/gh/privacyidea/privacyidea/pull/1040?src=pr&el=tree)
```diff
@@            Coverage Diff             @@
##           master    #1040      +/-   ##
==========================================
+ Coverage   95.41%   95.41%   +<.01%
==========================================
Files         131      131
Lines       16199    16203       +4
==========================================
+ Hits        15456    15460       +4
Misses        743      743
```
| [Impacted Files](https://codecov.io/gh/privacyidea/privacyidea/pull/1040?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [privacyidea/lib/utils.py](https://codecov.io/gh/privacyidea/privacyidea/pull/1040/diff?src=pr&el=tree#diff-cHJpdmFjeWlkZWEvbGliL3V0aWxzLnB5) | `96.69% <100%> (+0.01%)` | :arrow_up: |
| [privacyidea/lib/resolvers/SQLIdResolver.py](https://codecov.io/gh/privacyidea/privacyidea/pull/1040/diff?src=pr&el=tree#diff-cHJpdmFjeWlkZWEvbGliL3Jlc29sdmVycy9TUUxJZFJlc29sdmVyLnB5) | `97.44% <100%> (+0.01%)` | :arrow_up: |
------
[Continue to review full report at Codecov](https://codecov.io/gh/privacyidea/privacyidea/pull/1040?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/privacyidea/privacyidea/pull/1040?src=pr&el=footer). Last update [c8cb96c...6f5d4f1](https://codecov.io/gh/privacyidea/privacyidea/pull/1040?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).
- <a href='https://github.com/cornelinux'><img border=0 src='https://avatars1.githubusercontent.com/u/1908620?v=4' height=16 width=16></a> Thanks for the adaptions, they look good.
I just realized, that your PR is based on version 2.17. Would you mind rebasing it to the current master?
- <a href='https://github.com/andyboeh'><img border=0 src='https://avatars3.githubusercontent.com/u/3099753?v=4' height=16 width=16></a> I'm not very proficient in using git, but I know what you would like to achieve - my problem is that I started with master and not a feature branch and that I did a merge instead of a rebase in the first place..
- <a href='https://github.com/andyboeh'><img border=0 src='https://avatars3.githubusercontent.com/u/3099753?v=4' height=16 width=16></a> OK, I kind of fixed it locally, but if I push it now, I'm pretty sure it's going to destroy the PR, since I moved the relevant commits to a feature branch and cleaned-up my master branch. Should I create a new PR, referencing this one?
- <a href='https://github.com/cornelinux'><img border=0 src='https://avatars1.githubusercontent.com/u/1908620?v=4' height=16 width=16></a> Thanks for all your effort. I will probably take a look at it again tomorrow and maybe can fix this.
- [ ] <a href='#crh-comment-Pull 6f5d4f1e1f8a1372ddfbe10f9b599000a0b80c71 privacyidea/lib/utils.py 9'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/privacyidea/privacyidea/pull/1040#discussion_r186070706'>File: privacyidea/lib/utils.py:L849-868</a></b>
- <a href='https://github.com/cornelinux'><img border=0 src='https://avatars1.githubusercontent.com/u/1908620?v=4' height=16 width=16></a> To be able to choose these hashtypes (MD5CRYPT, SHA512CRYPT) in a configurable SQL Resolver, you need to add this to the WebUI dropdown here:
https://github.com/privacyidea/privacyidea/blob/master/privacyidea/static/components/config/controllers/configControllers.js#L990
- [ ] <a href='#crh-comment-Pull 6f5d4f1e1f8a1372ddfbe10f9b599000a0b80c71 privacyidea/lib/utils.py 8'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/privacyidea/privacyidea/pull/1040#discussion_r186113242'>File: privacyidea/lib/utils.py:L849-868</a></b>
- <a href='https://github.com/andyboeh'><img border=0 src='https://avatars3.githubusercontent.com/u/3099753?v=4' height=16 width=16></a> You are right, up to 16 bytes are supported by crypt(3), while I had exactly 8 bytes in mind. I'll double that.


<a href='https://www.codereviewhub.com/privacyidea/privacyidea/pull/1040?mark_as_completed=1'><img src='http://www.codereviewhub.com/site/github-mark-as-completed.png' height=26></a>&nbsp;<a href='https://www.codereviewhub.com/privacyidea/privacyidea/pull/1040?approve=1'><img src='http://www.codereviewhub.com/site/github-approve.png' height=26></a>&nbsp;<a href='https://github.com/privacyidea/privacyidea/pull/1040'><img src='http://www.codereviewhub.com/site/github-refresh.png' height=26></a>
<a href='#crh-end'></a>

cornelinux commented:
This pull request is now managed with CodeReviewHub. <a href='https://github.com/privacyidea/privacyidea/pull/1040'>Refresh</a>.<a href='#crh-update-none'></a>

cornelinux commented on privacyidea/lib/utils.py:
```diff
@@ -849,11 +849,19 @@ def hash_password(password, hashtype):
+        password = crypt.crypt(password, "$1$" + salt + "$")
+    elif hashtype == "SHA512CRYPT":
+        salt = geturandom(4, True)
+        password = crypt.crypt(password, "$6$" + salt + "$")
```
To be able to choose these hashtypes (MD5CRYPT, SHA512CRYPT) in a configurable SQL Resolver, you need to add this to the WebUI dropdown here:

https://github.com/privacyidea/privacyidea/blob/master/privacyidea/static/components/config/controllers/configControllers.js#L990

cornelinux commented on privacyidea/lib/utils.py:
```diff
@@ -849,11 +849,19 @@ def hash_password(password, hashtype):
+        salt = geturandom(4, True)
+        password = crypt.crypt(password, "$1$" + salt + "$")
+    elif hashtype == "SHA512CRYPT":
+        salt = geturandom(4, True)
```
Why wouldn't we add a bit more salt? (I like salty ;-)

andyboeh replied:
You are right, up to 16 bytes are supported by crypt(3), while I had exactly 8 bytes in mind. I'll double that.

cornelinux commented:
Thanks for the adaptions, they look good.
I just realized, that your PR is based on version 2.17. Would you mind rebasing it to the current master?

andyboeh commented:
I'm not very proficient in using git, but I know what you would like to achieve - my problem is that I started with master and not a feature branch and that I did a merge instead of a rebase in the first place..

andyboeh commented:
OK, I kind of fixed it locally, but if I push it now, I'm pretty sure it's going to destroy the PR, since I moved the relevant commits to a feature branch and cleaned-up my master branch. Should I create a new PR, referencing this one?

cornelinux commented:
Thanks for all your effort. I will probably take a look at it again tomorrow and maybe can fix this.
